### PR TITLE
protect instructiont::code

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/src/java_bytecode/remove_exceptions.cpp [hunk 1]
```diff
@@ -424,7 +423,7 @@ bool remove_exceptionst::instrument_throw(
     typecast_exprt(exc_expr, exc_thrown.type()));
   // now turn the `throw' into `assignment'
   instr_it->type=ASSIGN;
-  instr_it->code=assignment;
+  instr_it->code_nonconst() = assignment;
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/remove_exceptions.cpp:426]: Couldn't we instead replace the entire instruction here?

File: jbmc/src/java_bytecode/convert_java_nondet.cpp [hunk 2]
```diff
@@ -93,7 +93,7 @@ static std::pair<goto_programt::targett, bool> insert_nondet_init_code(
   // We only expect to find nondets in the rhs of an assignments, and in return
   // statements if remove_returns has not been run, but we do a more general
   // check on all operands in case this changes
-  for(exprt &op : target->code.operands())
+  for(exprt &op : target->code_nonconst().operands())
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/convert_java_nondet.cpp:96]: I believe this should be replaced by a call to `instructiont::transform`.

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 3]
```diff
@@ -244,8 +244,9 @@ bool remove_instanceoft::lower_instanceof(
   goto_programt &goto_program,
   goto_programt::targett target)
 {
-  if(target->is_target() &&
-     (contains_instanceof(target->code) || contains_instanceof(target->guard)))
+  if(
+    target->is_target() && (contains_instanceof(target->get_code()) ||
+                            contains_instanceof(target->guard)))
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/remove_instanceof.cpp:249]: I believe this should be handled via `instructiont::transform`.

File: jbmc/src/java_bytecode/replace_java_nondet.cpp [hunk 4]
```diff
@@ -282,8 +281,8 @@ static goto_programt::targett check_and_replace_target(
     inserted_expr.set_nullable(
       instr_info.get_nullable_type() ==
       nondet_instruction_infot::is_nullablet::TRUE);
-    target_instruction->code = code_returnt(inserted_expr);
-    target_instruction->code.add_source_location() =
+    target_instruction->code_nonconst() = code_returnt(inserted_expr);
+    target_instruction->code_nonconst().add_source_location() =
       target_instruction->source_location;
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/replace_java_nondet.cpp:286]: Why did we ever need to replace the entire `code` - wouldn't the `return_value()` suffice?

File: jbmc/src/java_bytecode/replace_java_nondet.cpp [hunk 5]
```diff
@@ -296,8 +295,9 @@ static goto_programt::targett check_and_replace_target(
     inserted_expr.set_nullable(
       instr_info.get_nullable_type() ==
       nondet_instruction_infot::is_nullablet::TRUE);
-    target_instruction->code = code_assignt(nondet_var, inserted_expr);
-    target_instruction->code.add_source_location() =
+    target_instruction->code_nonconst() =
+      code_assignt(nondet_var, inserted_expr);
+    target_instruction->code_nonconst().add_source_location() =
       target_instruction->source_location;
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/replace_java_nondet.cpp:301]: Just like the above: it's an `ASSIGN` instruction already, don't we just need to replace lhs/rhs here?

File: src/analyses/constant_propagator.cpp [hunk 6]
```diff
@@ -758,7 +758,7 @@ void constant_propagator_ait::replace(
     if(d.is_bottom())
       continue;
 
-    replace_types_rec(d.values.replace_const, it->code);
+    replace_types_rec(d.values.replace_const, it->code_nonconst());
```
Comment (tautschnig) [inline src/analyses/constant_propagator.cpp:761]: Could we do this via `instructiont::transform`?

File: src/analyses/goto_check.cpp [hunk 7]
```diff
@@ -2089,7 +2091,7 @@ void goto_checkt::goto_check(
           goto_programt::targett t =
             new_code.add(goto_programt::make_assignment(
               std::move(lhs), std::move(rhs), i.source_location));
-          t->code.add_source_location()=i.source_location;
+          t->code_nonconst().add_source_location() = i.source_location;
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp:2124]: This line should be removed and `goto_programt::make_assignment` should be fixed so as to invoke `code_assignt(std::move(lhs), std::move(rhs), l)` (i.e., adding the `, l`).

File: src/analyses/invariant_set_domain.cpp [hunk 8]
```diff
@@ -56,22 +56,22 @@ void invariant_set_domaint::transform(
...
     break;
 
   case DECL:
-    invariant_set.apply_code(from_l->code);
+    invariant_set.apply_code(from_l->get_code());
     break;
 
   case FUNCTION_CALL:
-    invariant_set.apply_code(from_l->code);
+    invariant_set.apply_code(from_l->get_code());
```
Comment (tautschnig) [inline src/analyses/invariant_set_domain.cpp:74]: `get_function_call()`

File: src/analyses/invariant_set_domain.cpp [hunk 9]
```diff
@@ -56,22 +56,22 @@ void invariant_set_domaint::transform(
...
 
   case OTHER:
     if(from_l->get_other().is_not_nil())
-      invariant_set.apply_code(from_l->code);
+      invariant_set.apply_code(from_l->get_code());
     break;
 
   case DECL:
-    invariant_set.apply_code(from_l->code);
+    invariant_set.apply_code(from_l->get_code());
```
Comment (tautschnig) [inline src/analyses/invariant_set_domain.cpp:70]: `get_decl()`

File: src/goto-cc/linker_script_merge.cpp [hunk 10]
```diff
@@ -261,8 +261,8 @@ int linker_script_merget::pointerize_linker_defined_symbols(
     goto_programt &program=gf.second.body;
     for(auto &instruction : program.instructions)
     {
-      for(exprt *insts :
-          std::list<exprt *>({&(instruction.code), &(instruction.guard)}))
+      for(exprt *insts : std::list<exprt *>(
+            {&(instruction.code_nonconst()), &(instruction.guard)}))
```
Comment (tautschnig) [inline src/goto-cc/linker_script_merge.cpp:265]: Probably a bit more work, but really should use `instructiont::transform`.

File: src/goto-instrument/accelerate/overflow_instrumenter.cpp [hunk 11]
```diff
@@ -54,7 +54,7 @@ void overflow_instrumentert::add_overflow_checks(
 
   if(t->is_assign())
   {
-    code_assignt &assignment=to_code_assign(t->code);
+    code_assignt &assignment = to_code_assign(t->code_nonconst());
```
Comment (tautschnig) [inline src/goto-instrument/accelerate/overflow_instrumenter.cpp:57]: I think this should use `get_assign()` followed by `set_assign()`

File: src/goto-instrument/accelerate/scratch_program.cpp [hunk 12]
```diff
@@ -133,7 +133,7 @@ void scratch_programt::fix_types()
   {
     if(it->is_assign())
     {
-      code_assignt &code=to_code_assign(it->code);
+      code_assignt &code = to_code_assign(it->code_nonconst());
```
Comment (tautschnig) [inline src/goto-instrument/accelerate/scratch_program.cpp:136]: Should use `get_assign()` followed by `set_assign()`

File: src/goto-instrument/concurrency.cpp [hunk 13]
```diff
@@ -108,7 +108,7 @@ void concurrency_instrumentationt::instrument(
   {
     if(it->is_assign())
     {
-      code_assignt &code=to_code_assign(it->code);
+      code_assignt &code = to_code_assign(it->code_nonconst());
```
Comment (tautschnig) [inline src/goto-instrument/concurrency.cpp:111]: Should use `get_assign()` followed by `set_assign()`.

File: src/analyses/invariant_set_domain.cpp [hunk 14]
```diff
@@ -56,22 +56,22 @@ void invariant_set_domaint::transform(
...
-      const code_assignt &assignment=to_code_assign(from_l->code);
+      const code_assignt &assignment = from_l->get_assign();
       invariant_set.assignment(assignment.lhs(), assignment.rhs());
     }
     break;
 
   case OTHER:
     if(from_l->get_other().is_not_nil())
-      invariant_set.apply_code(from_l->code);
+      invariant_set.apply_code(from_l->get_code());
```
Comment (tautschnig) [inline src/analyses/invariant_set_domain.cpp:66]: `get_other`

File: src/goto-instrument/concurrency.cpp [hunk 15]
```diff
@@ -119,7 +119,7 @@ void concurrency_instrumentationt::instrument(
     }
     else if(it->is_function_call())
     {
-      code_function_callt &code=to_code_function_call(it->code);
+      code_function_callt &code = to_code_function_call(it->code_nonconst());
```
Comment (tautschnig) [inline src/goto-instrument/concurrency.cpp:122]: `get_function_call`, `set_function_call()`

File: src/goto-instrument/generate_function_bodies.cpp [hunk 16]
```diff
@@ -625,7 +625,7 @@ void generate_function_bodies(
       if(instruction.is_function_call())
       {
         auto &called_function =
-          to_code_function_call(instruction.code).function();
+          to_code_function_call(instruction.code_nonconst()).function();
         if(is_havoc_function_call(called_function))
```
Comment (tautschnig) [inline src/goto-instrument/generate_function_bodies.cpp:629]: `get_function_call()`, `set_function_call()`

File: src/goto-instrument/loop_utils.cpp [hunk 17]
```diff
@@ -51,7 +51,7 @@ void build_havoc_code(
     goto_programt::targett t = dest.add(goto_programt::make_assignment(
       code_assignt(std::move(lhs), std::move(rhs)),
       loop_head->source_location));
-    t->code.add_source_location()=loop_head->source_location;
+    t->code_nonconst().add_source_location() = loop_head->source_location;
```
Comment (tautschnig) [inline src/goto-instrument/loop_utils.cpp:54]: As above: this line should be made unnecessary by fixing `goto_programt::make_assignment`.

File: src/goto-instrument/nondet_volatile.cpp [hunk 18]
```diff
@@ -226,16 +226,22 @@ void nondet_volatilet::nondet_volatile(
...
+        post);
     }
     else if(instruction.is_function_call())
     {
       // these have arguments and a return LHS
 
-      code_function_callt &code_function_call=
-        to_code_function_call(instruction.code);
+      code_function_callt &code_function_call =
+        to_code_function_call(instruction.code_nonconst());
```
Comment (tautschnig) [inline src/goto-instrument/nondet_volatile.cpp:244]: `get_function_call()`, `set_function_call()`

File: src/goto-instrument/uninitialized.cpp [hunk 19]
```diff
@@ -123,11 +123,11 @@ void uninitializedt::add_assertions(
         symbol_exprt symbol_expr(new_identifier, bool_typet());
         i1->type=DECL;
         i1->source_location=instruction.source_location;
-        i1->code=code_declt(symbol_expr);
+        i1->code_nonconst() = code_declt(symbol_expr);
```
Comment (tautschnig) [inline src/goto-instrument/uninitialized.cpp:126]: We should not be constructing an instruction in this way.

File: src/goto-instrument/uninitialized.cpp [hunk 20]
```diff
@@ -123,11 +123,11 @@ void uninitializedt::add_assertions(
...
 
         i2->type=ASSIGN;
         i2->source_location=instruction.source_location;
-        i2->code=code_assignt(symbol_expr, false_exprt());
+        i2->code_nonconst() = code_assignt(symbol_expr, false_exprt());
```
Comment (tautschnig) [inline src/goto-instrument/uninitialized.cpp:130]: We should not be constructing an instruction in this way.

File: src/goto-instrument/uninitialized.cpp [hunk 21]
```diff
@@ -182,7 +182,7 @@ void uninitializedt::add_assertions(
 
             goto_programt::instructiont assignment;
             assignment.type=ASSIGN;
-            assignment.code=code_assignt(
+            assignment.code_nonconst() = code_assignt(
```
Comment (tautschnig) [inline src/goto-instrument/uninitialized.cpp:185]: We should not be constructing an instruction in this way.

File: src/goto-instrument/value_set_fi_fp_removal.cpp [hunk 22]
```diff
@@ -125,13 +125,13 @@ void remove_function_pointer(
...
-    to_code_function_call(t1->code).function() = fun;
+    to_code_function_call(t1->code_nonconst()).function() = fun;
 
     // the signature of the function might not match precisely
     const namespacet ns(goto_model.symbol_table);
-    fix_argument_types(to_code_function_call(t1->code), ns);
+    fix_argument_types(to_code_function_call(t1->code_nonconst()), ns);
     fix_return_type(
-      to_code_function_call(t1->code), new_code_calls, goto_model);
+      to_code_function_call(t1->code_nonconst()), new_code_calls, goto_model);
```
Comment (tautschnig) [inline src/goto-instrument/value_set_fi_fp_removal.cpp:134]: All of the above should be handled via `set_function_call()`

File: src/goto-instrument/value_set_fi_fp_removal.cpp [hunk 23]
```diff
@@ -179,7 +179,7 @@ void remove_function_pointer(
   // further pointer-related errors.
   code_expressiont code_expression(function);
   code_expression.add_source_location() = function.source_location();
-  target->code.swap(code_expression);
+  target->code_nonconst().swap(code_expression);
   target->type = OTHER;
```
Comment (tautschnig) [inline src/goto-instrument/value_set_fi_fp_removal.cpp:183]: We should completely replace the instruction.

File: src/goto-instrument/wmm/shared_buffers.cpp [hunk 24]
```diff
@@ -176,8 +176,8 @@ void shared_bufferst::assignment(
 
     t=goto_program.insert_before(t);
     t->type=ASSIGN;
-    t->code=code_assignt(symbol, value);
-    t->code.add_source_location()=source_location;
+    t->code_nonconst() = code_assignt(symbol, value);
+    t->code_nonconst().add_source_location() = source_location;
```
Comment (tautschnig) [inline src/goto-instrument/wmm/shared_buffers.cpp:180]: We should not be constructing an instruction in this way.

File: src/goto-programs/goto_convert_exceptions.cpp [hunk 25]
```diff
@@ -113,7 +113,7 @@ void goto_convertt::convert_try_catch(
   // add the CATCH-pop to the end of the 'try' block
   goto_programt::targett catch_pop_instruction =
     dest.add(goto_programt::make_catch());
-  catch_pop_instruction->code = code_pop_catcht();
+  catch_pop_instruction->code_nonconst() = code_pop_catcht();
```
Comment (tautschnig) [inline src/goto-programs/goto_convert_exceptions.cpp:116]: It seems `make_catch()` needs to be fixed?

File: src/goto-programs/goto_inline_class.cpp [hunk 26]
```diff
@@ -56,7 +56,7 @@ void goto_inlinet::parameter_assignments(
 
     goto_programt::targett decl =
       dest.add(goto_programt::make_decl(symbol.symbol_expr(), source_location));
-    decl->code.add_source_location() = source_location;
+    decl->code_nonconst().add_source_location() = source_location;
```
Comment (tautschnig) [inline src/goto-programs/goto_inline_class.cpp:59]: `goto_programt::make_decl` needs to be fixed so as to set the source location.

File: src/goto-programs/goto_inline_class.cpp [hunk 27]
```diff
@@ -149,7 +149,7 @@ void goto_inlinet::parameter_destruction(
 
       goto_programt::targett dead = dest.add(
         goto_programt::make_dead(symbol.symbol_expr(), source_location));
-      dead->code.add_source_location()=source_location;
+      dead->code_nonconst().add_source_location() = source_location;
```
Comment (tautschnig) [inline src/goto-programs/goto_inline_class.cpp:152]: `make_dead` needs to be fixed so as to set the source location.

File: src/goto-programs/goto_inline_class.cpp [hunk 28]
```diff
@@ -169,7 +169,7 @@ void goto_inlinet::replace_return(
       {
         // a typecast may be necessary if the declared return type at the call
         // site differs from the defined return type
-        it->code = code_assignt(
+        it->code_nonconst() = code_assignt(
```
Comment (tautschnig) [inline src/goto-programs/goto_inline_class.cpp:172]: Completely replace the `instructiont`.

File: src/goto-programs/goto_inline_class.cpp [hunk 29]
```diff
@@ -288,7 +288,7 @@ void goto_inlinet::insert_function_body(
     for(auto &instruction : tmp.instructions)
     {
       replace_location(instruction.source_location, target->source_location);
-      replace_location(instruction.code, target->source_location);
+      replace_location(instruction.code_nonconst(), target->source_location);
```
Comment (tautschnig) [inline src/goto-programs/goto_inline_class.cpp:291]: Use `instructiont::transform`

File: src/goto-programs/goto_inline_class.cpp [hunk 30]
```diff
@@ -301,7 +301,7 @@ void goto_inlinet::insert_function_body(
 
   // kill call
   target->type=LOCATION;
-  target->code.clear();
+  target->code_nonconst().clear();
```
Comment (tautschnig) [inline src/goto-programs/goto_inline_class.cpp:304]: Completely replace the instruction.

File: src/goto-programs/label_function_pointer_call_sites.cpp [hunk 31]
```diff
@@ -61,7 +61,7 @@ void label_function_pointer_call_sites(goto_modelt &goto_model)
 
         goto_function.second.body.insert_before_swap(it, assign_instruction);
         const auto next = std::next(it);
-        to_code_function_call(next->code).function() =
+        to_code_function_call(next->code_nonconst()).function() =
```
Comment (tautschnig) [inline src/goto-programs/label_function_pointer_call_sites.cpp:64]: Use `set_function_call`

File: src/goto-programs/link_goto_model.cpp [hunk 32]
```diff
@@ -36,7 +36,7 @@ static void rename_symbols_in_function(
 
   for(auto &instruction : function.body.instructions)
   {
-    rename_symbol(instruction.code);
+    rename_symbol(instruction.code_nonconst());
```
Comment (tautschnig) [inline src/goto-programs/link_goto_model.cpp:39]: Use `instructiont::transform`

File: src/goto-programs/string_abstraction.cpp [hunk 33]
```diff
@@ -1293,23 +1295,23 @@ goto_programt::targett string_abstractiont::value_assignments_string_struct(
...
+    assignment->code_nonconst().add_source_location() = target->source_location;
   }
 
   {
     goto_programt::targett assignment = tmp.add(goto_programt::make_assignment(
       member(lhs, whatt::SIZE),
       member(rhs, whatt::SIZE),
       target->source_location));
-    assignment->code.add_source_location()=target->source_location;
+    assignment->code_nonconst().add_source_location() = target->source_location;
```
Comment (tautschnig) [inline src/goto-programs/string_abstraction.cpp:1355]: This line should not be necessary.

File: src/goto-programs/string_abstraction.cpp [hunk 34]
```diff
@@ -1293,23 +1295,23 @@ goto_programt::targett string_abstractiont::value_assignments_string_struct(
...
   }
 
   {
     goto_programt::targett assignment = tmp.add(goto_programt::make_assignment(
       member(lhs, whatt::LENGTH),
       member(rhs, whatt::LENGTH),
       target->source_location));
-    assignment->code.add_source_location()=target->source_location;
+    assignment->code_nonconst().add_source_location() = target->source_location;
```
Comment (tautschnig) [inline src/goto-programs/string_abstraction.cpp:1347]: This line should not be necessary.

File: src/goto-programs/string_instrumentation.cpp [hunk 35]
```diff
@@ -882,5 +882,5 @@ void string_instrumentationt::invalidate_buffer(
   const side_effect_expr_nondett nondet(
     buf_type.subtype(), target->source_location);
 
-  invalidate->code=code_assignt(deref, nondet);
+  invalidate->code_nonconst() = code_assignt(deref, nondet);
```
Comment (tautschnig) [inline src/goto-programs/string_instrumentation.cpp:887]: Use `set_assign`

**Review by tautschnig — COMMENTED** [review]
I am aware that quite a few of my comments amount to shooting-the-messenger, but I really think we need to use this opportunity to fix the code.

**Comment by kroening** [comment]
The suggested changes are all agreed -- but should come as a separate PR, to keep this reviewable.

**Review by tautschnig — APPROVED** [review]
Approving under the assumption that the follow-up PR will be created that cleans up most (all?) uses of `code_nonconst()` as suggested in my earlier comments.

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**